pkg/epp/scheduling/scheduler.go

```diff
@@ -20,141 +20,197 @@ package scheduling
+				Current: plugins.LeastQueueFilter,
+				NextOnSuccessOrFailure: &plugins.DecisionTreeFilter{
+					Current: plugins.LeastKVCacheFilter,
 				},
 			},
 		},
-		nextOnFailure: &decisionTreeFilter{
-			current: leastQueueFilter,
-			nextOnSuccessOrFailure: &decisionTreeFilter{
-				current: loRAAffinityFilter,
-				nextOnSuccessOrFailure: &decisionTreeFilter{
-					current: leastKVCacheFilter,
+		NextOnFailure: &plugins.DecisionTreeFilter{
+			Current: plugins.LeastQueueFilter,
+			NextOnSuccessOrFailure: &plugins.DecisionTreeFilter{
+				Current: plugins.LoRAAffinityFilter,
+				NextOnSuccessOrFailure: &plugins.DecisionTreeFilter{
+					Current: plugins.LeastKVCacheFilter,
 				},
 			},
 		},
 	}
 
-	sheddableRequestFilter = &decisionTreeFilter{
+	sheddableRequestFilter = &plugins.DecisionTreeFilter{
 		// When there is at least one model server that's not queuing requests, and still has KV
 		// cache below a certain threshold, we consider this model server has capacity to handle
 		// a sheddable request without impacting critical requests.
-		current:       hasCapacityFilter,
-		nextOnSuccess: lowLatencyFilter,
+		Current:       plugins.HasCapacityFilter,
+		NextOnSuccess: lowLatencyFilter,
 		// If all pods are queuing or running above the KVCache threshold, we drop the sheddable
 		// request to make room for critical requests.
-		nextOnFailure: dropRequestFilter,
-	}
-
-	hasCapacityFilter = &basicFilter{
-		name:   "has capacity for sheddable requests",
-		filter: toFilterFunc(queueThresholdPredicate(config.QueueThresholdCritical).and(kvCacheThresholdPredicate(config.KVCacheThreshold))),
-	}
-
-	dropRequestFilter = &basicFilter{
-		name: "drop request",
-		filter: func(ctx *types.Context, pods []*types.PodMetrics) ([]*types.PodMetrics, error) {
-			ctx.Logger.V(logutil.DEFAULT).Info("Request dropped", "request", ctx.Req)
-			return []*types.PodMetrics{}, errutil.Error{
-				Code: errutil.InferencePoolResourceExhausted, Msg: "dropping request due to limited backend resources",
-			}
-		},
+		NextOnFailure: plugins.DropRequestFilter,
 	}
 )
 
 func NewScheduler(datastore Datastore) *Scheduler {
+	defaultPlugin := &defaultPlugin{}
+
 	return &Scheduler{
-		datastore:              datastore,
-		criticalRequestFilter:  lowLatencyFilter,
-		sheddableRequestFilter: sheddableRequestFilter,
+		datastore:           datastore,
+		preSchedulePlugins:  []types.PreSchedule{},
+		postSchedulePlugins: []types.PostSchedule{},
+		scorers:             []types.Scorer{},
+		filters:             []types.Filter{defaultPlugin},
+		picker:              defaultPlugin,
 	}
 }
 
 type Scheduler struct {
-	datastore              Datastore
-	criticalRequestFilter  Filter
-	sheddableRequestFilter Filter
+	datastore           Datastore
+	preSchedulePlugins  []types.PreSchedule
+	postSchedulePlugins []types.PostSchedule
+	filters             []types.Filter
+	scorers             []types.Scorer
+	picker              types.Picker
```

Comment:
I'm reading this again, and Picker interface seems redundant to me.
the scheduler itself IS the picker.
scheduler runs filters to remove pods, scorers to score the filtered pods, and then pick a pod from the result list based on the calculated score of each.
maybe it would make sense to put something like SchedulerStrategy. I can image two possible strategies:
- pick the highest, fallback to second, etc.
- pick randomly while using the score for probabilities.
in current code the only option is random and therefore I wouldn't do the above suggestion in this PR, but only when (and if) needed.